state-sync/state-sync-v2/state-sync-driver/src/notification_handlers.rs

```diff
@@ -0,0 +1,328 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    driver_client::{ClientNotificationListener, DriverNotification},
+    error::Error,
+    storage_synchronizer::StorageStateSummary,
+};
+use consensus_notifications::{
+    ConsensusCommitNotification, ConsensusNotification, ConsensusNotificationListener,
+    ConsensusSyncNotification,
+};
+use diem_types::transaction::Transaction;
+use futures::{
+    channel::{mpsc, oneshot},
+    stream::FusedStream,
+    Stream,
+};
+use mempool_notifications::MempoolNotificationSender;
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+    time::{Duration, SystemTime},
+};
+
+// TODO(joshlind): make these configurable!
+const CONSENSUS_SYNC_REQUEST_TIMEOUT_MS: u64 = 120000; // 2 minutes
+const MEMPOOL_COMMIT_ACK_TIMEOUT_MS: u64 = 5000; // 5 seconds
+
+/// A simple handler for sending bootstrap notifications to client listeners
+pub struct BootstrapNotificationHandler {
+    // If the node has has already been bootstrapped
+    bootstrapped: bool,
+
+    // The listener for bootstrap requests from clients
+    client_listener: ClientNotificationListener,
+
+    // The channel used to notify a listener of successful bootstrapping
+    notifier_channel: Option<oneshot::Sender<Result<(), Error>>>,
+}
+
+impl BootstrapNotificationHandler {
+    pub fn new(client_notifications: mpsc::UnboundedReceiver<DriverNotification>) -> Self {
+        let client_listener = ClientNotificationListener::new(client_notifications);
+
+        Self {
+            bootstrapped: false,
+            client_listener,
+            notifier_channel: None,
+        }
+    }
+
+    /// Notifies the channel once the node has completed bootstrapping
+    pub fn notify_once_bootstrapped(
+        &mut self,
+        notifier_channel: oneshot::Sender<Result<(), Error>>,
+    ) {
+        self.notifier_channel = Some(notifier_channel);
+    }
```

Comment:
It doesn't send anything here. Instead it sets the notifier and then `notify_bootstrapped` will be called by state sync when the node has eventually bootstrapped.

Reply:
It's just weird to have a function called `notify_once_bootstrapped` and it doesn't notify.

Maybe,` add_bootstrap_notifier`?

Reply:
Cool, done! Note: I don't feel too strongly about a lot of this stuff at the moment because most of it will change. I just need to get a basic skeleton in so I can start iterating on the logic 😄